src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+        assert(0);
+        return 0;
+    }
+
+    // Verify our source port.
+    if (sport != zconf.target_port) {
+        return 0;
+    }
+
+    // Verify our packet length.
+    uint16_t udp_len = ntohs(udp->uh_ulen);
+    if (udp_len < dns_packet_len) {
+        return 0;
+    }
+
+    // Verify the packet length is ok.
+    if (len < udp_len) {
+        return 0;
+    }
+
+    // Looks good.
+    return 1;
+}
+
+void dns_process_packet(const u_char *packet, uint32_t len, fieldset_t *fs,
+        uint32_t *validation) 
+{   
+    struct ip *ip_hdr = (struct ip *) &packet[sizeof(struct ether_header)];
+
+    log_trace("dns", "processing packet from %s len %u", make_ip_str(ip_hdr->ip_src.s_addr), len);
+
+    //fs_add_string(fs, "icmp_responder", make_ip_str(ip_hdr->ip_src.s_addr), 1);
+    if (ip_hdr->ip_p == IPPROTO_UDP) {
+
+        struct udphdr *udp_hdr = (struct udphdr *) ((char *) ip_hdr + ip_hdr->ip_hl * 4);
+        uint16_t udp_len = ntohs(udp_hdr->uh_ulen);
+
+        // Handles in validate.
+        assert(udp_len >= dns_packet_len); 
+
+        char* qname_p = NULL;
+        dns_question_tail* tail_p = NULL;
+        bool is_valid = 0;
+        
+        dns_header* dns_header_p = (dns_header*) &udp_hdr[1];
+     
+        // verify our dns transaction id
+        if (dns_header_p->id == (validation[2] & 0xFFFF)) {
+
+            // Verify our question
+            qname_p = (char*) dns_header_p + sizeof(dns_header);
+            tail_p = (dns_question_tail*)(dns_packet + sizeof(dns_header) + qname_len);
+
+            // Verify our qname
+            if (strcmp(qname, qname_p) == 0) {
+            
+                // Verify the qtype and qclass.
+                if (tail_p->qtype == htons(qtype) && tail_p->qclass == htons(0x01)) {
+                    is_valid = 1;
+                }
+            }
+        }
+    
+        dns_header* dns_hdr = (dns_header*) &udp_hdr[1];
+        
+        uint16_t qr = dns_hdr->qr;
+        uint16_t rcode = dns_hdr->rcode;
+
+        // Success: Has the right validation bits and the right Q
+        // App success: has qr and rcode bits right
+        // Any app level parsing issues: dns_parse_err
+
+        // High level info
+        fs_add_string(fs, "classification", (char*) "dns", 0);
+        fs_add_uint64(fs, "success", is_valid);
+        fs_add_uint64(fs, "app_success", is_valid && (qr == DNS_QR_ANSWER) && (rcode == DNS_RCODE_NOERR));
+        
+        // UDP info
+        fs_add_uint64(fs, "udp_sport", ntohs(udp_hdr->uh_sport));
+        fs_add_uint64(fs, "udp_dport", ntohs(udp_hdr->uh_dport));
+        fs_add_uint64(fs, "udp_len", udp_len);
+        
+        // ICMP info
+        fs_add_null(fs, "icmp_responder");
+        fs_add_null(fs, "icmp_type");
+        fs_add_null(fs, "icmp_code");
+        fs_add_null(fs, "icmp_unreach_str");
+        
+        if (!is_valid) {
+            // DNS header
+            fs_add_null(fs, "dns_id"); 
```

Comment:
because this is going to be JSON, putting dns_* in front of every field will actually lead to a negligible increase in file sizes. Like 2.5 GB worth for a scan that returns 30 million responses. This will end up resolving itself when the file is compressed, but we should consider whether they're actually necessary.